### PR TITLE
Testing

### PR DESCRIPTION
**Pull request by lilian-hunt**
Implement the FindMeetingQuery method

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -14,10 +14,68 @@
...
+import java.util.List;
+import java.util.Set;
+import java.util.Timer;
+import java.util.stream.Collectors;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    long duration = request.getDuration();
+    // The duration should not be longer than 1 day
```
Comment (KathleenQ) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: A dot at the end of comment?

**Review by KathleenQ — COMMENTED** [review]
The codes look good to me.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,68 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+import java.util.Timer;
+import java.util.stream.Collectors;
```
Comment (aannie-y) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:23]: very minor, but imports should be in alphabetical order as per [this](https://google.github.io/styleguide/javaguide.html#s3.3.3-import-ordering-and-spacing)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/servlets/GetEventsServlet.java [hunk 3]
```diff
@@ -14,8 +14,8 @@
 
 package com.google.sps.servlets;
 
-import com.google.sps.Events;
 import com.google.gson.Gson;
+import com.google.sps.Events;
```
Comment (aannie-y) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/servlets/GetEventsServlet.java:18]: Same as above, have the import sorted

File: walkthroughs/week-5-tdd/project/src/main/webapp/script.js [hunk 4]
```diff
@@ -24,14 +24,16 @@ function sendMeetingRequest() {
   // split it into an array of names
   const attendees = attendeesNamesString.split(/\s*,\s*/);
 
-// comma-separated list of names
-  const optionalAttendeesNamesString = document.getElementById('optional-attendees').value;
+  // comma-separated list of names
+  const optionalAttendeesNamesString =
+      document.getElementById('optional-attendees').value;
```
Comment (aannie-y) [inline walkthroughs/week-5-tdd/project/src/main/webapp/script.js:29]: indentation should be 2 spaces

File: walkthroughs/week-5-tdd/project/src/main/webapp/script.js [hunk 5]
```diff
@@ -24,14 +24,16 @@ function sendMeetingRequest() {
...
   // split it into an array of names
   const optionalAttendees = optionalAttendeesNamesString.split(/\s*,\s*/);
 
   // Create the request to send to the server using the data we collected from
   // the web form.
-  const meetingRequest = new MeetingRequest(duration, attendees, optionalAttendees);
+  const meetingRequest =
+      new MeetingRequest(duration, attendees, optionalAttendees);
```
Comment (aannie-y) [inline walkthroughs/week-5-tdd/project/src/main/webapp/script.js:36]: indentation again

**Review by aannie-y — COMMENTED** [review]
Some minor indentation and import sorting but other than that LGTM in terms of style.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (lilian-hunt) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Added, thanks

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/servlets/GetEventsServlet.java (hunk 3 above)
Comment (lilian-hunt) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/servlets/GetEventsServlet.java:18]: Ok yep just ran clang format again

File: walkthroughs/week-5-tdd/project/src/main/webapp/script.js (hunk 5 above)
Comment (lilian-hunt) [reply walkthroughs/week-5-tdd/project/src/main/webapp/script.js:36]: Ah clang-format-path does not run on js files, I just realised. Trying to find another tool that does now

File: walkthroughs/week-5-tdd/project/src/main/webapp/script.js [hunk 6]
```diff
@@ -24,14 +24,16 @@ function sendMeetingRequest() {
   // split it into an array of names
   const attendees = attendeesNamesString.split(/\s*,\s*/);
 
-// comma-separated list of names
-  const optionalAttendeesNamesString = document.getElementById('optional-attendees').value;
+  // comma-separated list of names
```
Comment (jupi3200) [inline walkthroughs/week-5-tdd/project/src/main/webapp/script.js:27]: How is this change related to the Pull Request?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -14,10 +14,68 @@
...
+    int initialStart = TimeRange.START_OF_DAY;
+
+    // Find a gap greater than or equal to the duration.
+    for (TimeRange timeRange : mergedBusyTimes) {
+      System.out.println(timeRange.start() - initialStart);
+      if (timeRange.start() - initialStart >= duration) {
+        availableTimes.add(TimeRange.fromStartEnd(initialStart, timeRange.start(), false));
+      }
+      initialStart = timeRange.end();
+      System.out.println(timeRange.toString());
```
Comment (jupi3200) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Please avoid using System.out, prefer java.util.logging (This is more useful for server development as the logs could be directed to the correct file).

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 8]
```diff
@@ -14,10 +14,68 @@
...
+                                     .filter(event.getAttendees()::contains)
+                                     .collect(Collectors.toSet());
+      if (intersection.size() > 0) {
+        busyTimes.add(event.getWhen());
+      }
+    }
+
+    busyTimes.sort(TimeRange.ORDER_BY_START);
+    List<TimeRange> mergedBusyTimes = new ArrayList<>();
+    for (TimeRange timeRange : busyTimes) {
```
Comment (jupi3200) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: how does naming the variable _timeRange_ helping us understand the code?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 9]
```diff
@@ -14,10 +14,68 @@
...
+      return Arrays.asList();
+    }
+    Collection<String> attendees = request.getAttendees();
+    List<TimeRange> busyTimes = new ArrayList<>();
+
+    // Find the other events the attendees are already scheduled in for.
+    for (Event event : events) {
+      Set<String> intersection = attendees.stream()
+                                     .distinct()
+                                     .filter(event.getAttendees()::contains)
```
Comment (jupi3200) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#anyMatch-java.util.function.Predicate- might be idiomatic way to solve this.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 10]
```diff
@@ -14,10 +14,68 @@
...
+      // If the merged list is empty or the new interval does not overlap, append it.
+      // Otherwise, there must be overlap so we create a new TimeRange for the merged intervals.
+      if (mergedBusyTimes.size() == 0
+          || !(mergedBusyTimes.get(mergedBusyTimes.size() - 1)).overlaps(timeRange)) {
+        mergedBusyTimes.add(timeRange);
+      } else {
+        TimeRange lastFromMergedTimes = mergedBusyTimes.remove(mergedBusyTimes.size() - 1);
+        TimeRange mergedRange = TimeRange.fromStartEnd(lastFromMergedTimes.start(),
+            Math.max(lastFromMergedTimes.end(), timeRange.end()), false);
+        mergedBusyTimes.add(mergedRange);
```
Comment (jupi3200) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: How much time would it take to understand this code? Can it be written in a way that would improve the readability?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/MeetingRequest.java [hunk 11]
```diff
@@ -21,7 +21,6 @@
 import java.util.HashSet;
 
 public final class MeetingRequest {
-
```
Comment (jupi3200) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/MeetingRequest.java:24]: Why was this file modified?

Comment (lilian-hunt) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/MeetingRequest.java:24]: I ran clang format, so it removed the extra white line.

File: walkthroughs/week-5-tdd/project/src/main/webapp/script.js (hunk 4 above)
Comment (lilian-hunt) [reply walkthroughs/week-5-tdd/project/src/main/webapp/script.js:29]: Actually clang-format does work here, and the reason it changed this is because the line length was too long.

Comment (lilian-hunt) [reply walkthroughs/week-5-tdd/project/src/main/webapp/script.js:29]: So this is in fact correct.

File: walkthroughs/week-5-tdd/project/src/main/webapp/script.js (hunk 6 above)
Comment (lilian-hunt) [reply walkthroughs/week-5-tdd/project/src/main/webapp/script.js:27]: This was just a formatting change from running clang-format-path because the line was too long.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 7 above)
Comment (lilian-hunt) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Sorry, this was just from my testing purpose to see how the function worked. It is not necessary and has been deleted.

File: walkthroughs/week-5-tdd/project/src/main/webapp/script.js (hunk 4 above)
Comment (lilian-hunt) [reply walkthroughs/week-5-tdd/project/src/main/webapp/script.js:29]: As below, this is a result of clang-format because the line length was too long.

File: walkthroughs/week-5-tdd/project/src/main/webapp/script.js (hunk 5 above)
Comment (lilian-hunt) [reply walkthroughs/week-5-tdd/project/src/main/webapp/script.js:36]: In fact this was clang-format, it does work on JS files. It was because the line length was too long.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 8 above)
Comment (lilian-hunt) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Changed the name to unavailableTimeRange to help clarity

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 9 above)
Comment (lilian-hunt) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Thanks for this! Now using: 
attendees.stream().anyMatch(x -> event.getAttendees().contains(x))
to check which is more memory efficient.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 10 above)
Comment (lilian-hunt) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I have now split it up into variables for startTime and endTime to make it clearer how the start/end times for the interval are calculated.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 8 above)
Comment (jupi3200) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Thanks

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 9 above)
Comment (jupi3200) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Thanks

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 10 above)
Comment (jupi3200) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: :-)

**Review by jupi3200 — APPROVED** [review]
